Use the relative sysroot path in the linker command line to specify sysroot rlibs

danakj opened the pull request:
This addresses https://github.com/rust-lang/rust/issues/112586

danakj commented:
FWIW I considered putting the `FilePathMapping` object into each FooLinker as a member, however then it makes borrowing harder as it holds a borrow on `self` in the return and you end up having to copy the returned Path every time. This is worse in the (normal) case of no remappings, where the `FilePathMapping` is an empty Vec and thus creating it as a temporary is ~free. So I went back to just calling it through the `sess.opts`.

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -329,7 +329,8 @@ impl<'a> GccLinker<'a> {
                 let mut rpath = OsString::from("@rpath/");
-                rpath.push(out_filename.file_name().unwrap());
+                let mapping = self.sess.opts.file_path_mapping();
+                rpath.push(mapping.map_prefix(out_filename).0.file_name().unwrap());
```
This is already just the filename and not the full path, so `-Zremap-cwd-prefix` doesn't have any effect and I'm not sure `--remap-path-prefix` is supposed to allow remapping individual files as opposed to directories. It probably doesn't break anythinf though.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5260744340/jobs/9508063560) [(plain)](https://github.com/rust-lang/rust/commit/e7321b92b54a145ea105e752b1555f13c133b450/checks/14237611172/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/remap_path_prefix/xcrate-generic.rs stdout ----

error: auxiliary build of "/checkout/tests/codegen/remap_path_prefix/auxiliary/xcrate-generic.rs" failed to compile: 
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen/remap_path_prefix/auxiliary/xcrate-generic.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-O" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/xcrate-generic/auxiliary" "-Crpath" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-g" "--remap-path-prefix=/checkout/obj=/the/aux-cwd" "--remap-path-prefix=/checkout/tests/codegen/remap_path_prefix/auxiliary=/the/aux-src" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/xcrate-generic/auxiliary"
stdout: none
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: LC_ALL="C" PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" VSLANG="1033" "cc" "-Wl,--version-script=/tmp/rustcAAkMnw/list" "-Wl,--no-undefined-version" "-m64" "/tmp/rustcAAkMnw/symbols.o" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/xcrate-generic/auxiliary/xcrate-generic.xcrate_generic.af543f0758af3f22-cgu.0.rcgu.o" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/xcrate-generic/auxiliary/xcrate-generic.hoi31sisk3todsd.rcgu.rmeta" "-Wl,--as-needed" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/xcrate-generic/auxiliary" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bdynamic" "-lstd-9ae48462c8a66b65" "-Wl,-Bstatic" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-793f118aad406fce.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/xcrate-generic/auxiliary/libxcrate_generic.so" "-shared" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--enable-new-dtags" "-Wl,-z,origin"
  = note: /usr/bin/ld: cannot open output file /the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/xcrate-generic/auxiliary/libxcrate_generic.so: No such file or directory
          collect2: error: ld returned 1 exit status

error: aborting due to previous error
------------------------------------------


Build completed unsuccessfully in 0:12:48

---- [codegen] tests/codegen/remap_path_prefix/main.rs stdout ----

error: auxiliary build of "/checkout/tests/codegen/remap_path_prefix/auxiliary/remap_path_prefix_aux.rs" failed to compile: 
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen/remap_path_prefix/auxiliary/remap_path_prefix_aux.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-O" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main/auxiliary" "-Crpath" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-g" "--remap-path-prefix=/checkout/obj=/the/aux-cwd" "--remap-path-prefix=/checkout/tests/codegen/remap_path_prefix/auxiliary=/the/aux-src" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main/auxiliary"
stdout: none
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: LC_ALL="C" PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" VSLANG="1033" "cc" "-Wl,--version-script=/tmp/rustclKRoql/list" "-Wl,--no-undefined-version" "-m64" "/tmp/rustclKRoql/symbols.o" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main/auxiliary/remap_path_prefix_aux.remap_path_prefix_aux.c9e8e2bc76fc5d8f-cgu.0.rcgu.o" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main/auxiliary/remap_path_prefix_aux.2w9qdqd6o8rmtmh3.rcgu.rmeta" "-Wl,--as-needed" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main/auxiliary" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bdynamic" "-lstd-9ae48462c8a66b65" "-Wl,-Bstatic" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-793f118aad406fce.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main/auxiliary/libremap_path_prefix_aux.so" "-shared" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--enable-new-dtags" "-Wl,-z,origin"
  = note: /usr/bin/ld: cannot open output file /the/aux-cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main/auxiliary/libremap_path_prefix_aux.so: No such file or directory
          collect2: error: ld returned 1 exit status

error: aborting due to previous error
------------------------------------------

```

</details>

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -461,19 +466,19 @@ impl<'a> Linker for GccLinker<'a> {
     fn link_rlib(&mut self, lib: &Path) {
         self.hint_static();
-        self.cmd.arg(lib);
+        self.cmd.arg(self.sess.opts.file_path_mapping().map_prefix(lib).0.as_ref());
```
If the remap is anything other than turning an absolute path into a relative path that still resolves to the same absolute path given the working directory of rustc this will break linking as the linker actually needs to read the file at this path. So for example remapping `/path/to/rust_source` to `/rustc/$hash` as is done by rust's CI will be broken.

This is the reason for the CI failure.

michaelwoerister commented:
I can take the review for this.
r? @michaelwoerister

danakj commented:
Ah ok so there's 2 issues here which I will address. Thanks for the feedback so far.

1. We need to differentiate between remap-path-prefix and remap-cwd-prefix, where the latter is allowed to be applied to paths used for IO and thus must be a valid path still. I have an idea of how to express that in the code.
2. I understand your point about the linker impls being the wrong place @petrochenkov, though the upside to being there is that we can be sure it happens, whereas finding and ensuring all _callers_ do it is a bit harder. What if we replace &Path with a newtype MappedPath(&'a Path) for the methods that show up in the linker output (rlib and object paths) and this way help ensure callers do the mapping?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5266956272/jobs/9521581694) [(plain)](https://github.com/rust-lang/rust/commit/52c9394f633c70766e263ec1a0a1a7bc25352091/checks/14254757901/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking quote v1.0.26
    Checking rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
    Checking syn v2.0.8
error[E0277]: can't compare `(PathBuf, PathBuf, RemapForIO)` with `(_, _)`
    |
86  |   macro_rules! assert_ne {
    |   ---------------------- in this expansion of `assert_ne!` (#2)
...
...
90  |                   if *left_val == *right_val {
    |                                ^^ no implementation for `(PathBuf, PathBuf, RemapForIO) == (_, _)`
   ::: compiler/rustc_interface/src/tests.rs:623:5
    |
623 | /     macro_rules! tracked {
623 | /     macro_rules! tracked {
624 | |         ($name: ident, $non_default_value: expr) => {
625 | |             opts = reference.clone();
626 | |             assert_ne!(opts.$name, $non_default_value);
...   |
632 | |         };
633 | |     }
    | |_____- in this expansion of `tracked!` (#1)
    | |_____- in this expansion of `tracked!` (#1)
...
641 |       tracked!(remap_path_prefix, vec![("/home/bors/rust".into(), "src".into())]);
    |
    |
    = help: the trait `std::cmp::PartialEq<(_, _)>` is not implemented for `(PathBuf, PathBuf, RemapForIO)`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              ()
              (A, Z, Y, X, W, V, U, T)
              (B, A, Z, Y, X, W, V, U, T)
              (C, B, A, Z, Y, X, W, V, U, T)
              (D, C, B, A, Z, Y, X, W, V, U, T)
              (E, D, C, B, A, Z, Y, X, W, V, U, T)
              (T,)
              (U, T)
            and 5 others
    = note: required for `Vec<(PathBuf, PathBuf, RemapForIO)>` to implement `std::cmp::PartialEq<Vec<(_, _)>>`
    Checking rustc_metadata v0.0.0 (/checkout/compiler/rustc_metadata)
    Checking rustc_metadata v0.0.0 (/checkout/compiler/rustc_metadata)
error[E0277]: `RemapForIO` doesn't implement `std::fmt::Debug`
    |
86  |   macro_rules! assert_ne {
    |   ---------------------- in this expansion of `assert_ne!` (#2)
...
...
95  |                       $crate::panicking::assert_failed(kind, &*left_val, &*right_val, $crate::option::Option::None);
    |                                                              ^^^^^^^^^^ `RemapForIO` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
   ::: compiler/rustc_interface/src/tests.rs:623:5
    |
623 | /     macro_rules! tracked {
623 | /     macro_rules! tracked {
624 | |         ($name: ident, $non_default_value: expr) => {
625 | |             opts = reference.clone();
626 | |             assert_ne!(opts.$name, $non_default_value);
...   |
632 | |         };
633 | |     }
    | |_____- in this expansion of `tracked!` (#1)
    | |_____- in this expansion of `tracked!` (#1)
...
641 |       tracked!(remap_path_prefix, vec![("/home/bors/rust".into(), "src".into())]);
    |
    = help: the trait `std::fmt::Debug` is not implemented for `RemapForIO`
    = help: the trait `std::fmt::Debug` is implemented for `Vec<T, A>`


    Checking rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
error[E0308]: mismatched types
   --> compiler/rustc_interface/src/tests.rs:641:38
    |
641 |     tracked!(remap_path_prefix, vec![("/home/bors/rust".into(), "src".into())]);
    |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected a tuple with 3 elements, found one with 2 elements
    |
    = note: expected tuple `(PathBuf, PathBuf, RemapForIO)`
               found tuple `(PathBuf, PathBuf)`
    Checking rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `rustc_interface` (lib test) due to 3 previous errors
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5267022468/jobs/9521734611) [(plain)](https://github.com/rust-lang/rust/commit/f5a9805943ccafbef3224bd8a65722255b6c440c/checks/14254955857/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
    Checking rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
    Checking syn v2.0.8
error[E0277]: `RemapForIO` doesn't implement `std::fmt::Debug`
    |
86  |   macro_rules! assert_ne {
    |   ---------------------- in this expansion of `assert_ne!` (#2)
...
...
95  |                       $crate::panicking::assert_failed(kind, &*left_val, &*right_val, $crate::option::Option::None);
    |                                                              ^^^^^^^^^^ `RemapForIO` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
   ::: compiler/rustc_interface/src/tests.rs:624:5
    |
624 | /     macro_rules! tracked {
624 | /     macro_rules! tracked {
625 | |         ($name: ident, $non_default_value: expr) => {
626 | |             opts = reference.clone();
627 | |             assert_ne!(opts.$name, $non_default_value);
...   |
633 | |         };
634 | |     }
    | |_____- in this expansion of `tracked!` (#1)
    | |_____- in this expansion of `tracked!` (#1)
...
642 |       tracked!(remap_path_prefix, vec![("/home/bors/rust".into(), "src".into(), RemapForIO::No)]);
    |
    = help: the trait `std::fmt::Debug` is not implemented for `RemapForIO`
    = help: the trait `std::fmt::Debug` is implemented for `Vec<T, A>`


error[E0277]: `RemapForIO` doesn't implement `std::fmt::Debug`
    |
86  |   macro_rules! assert_ne {
    |   ---------------------- in this expansion of `assert_ne!` (#2)
...
...
95  |                       $crate::panicking::assert_failed(kind, &*left_val, &*right_val, $crate::option::Option::None);
    |                                                                          ^^^^^^^^^^^ `RemapForIO` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
   ::: compiler/rustc_interface/src/tests.rs:624:5
    |
624 | /     macro_rules! tracked {
624 | /     macro_rules! tracked {
625 | |         ($name: ident, $non_default_value: expr) => {
626 | |             opts = reference.clone();
627 | |             assert_ne!(opts.$name, $non_default_value);
...   |
633 | |         };
634 | |     }
    | |_____- in this expansion of `tracked!` (#1)
    | |_____- in this expansion of `tracked!` (#1)
...
642 |       tracked!(remap_path_prefix, vec![("/home/bors/rust".into(), "src".into(), RemapForIO::No)]);
    |
    = help: the trait `std::fmt::Debug` is not implemented for `RemapForIO`
    = help: the trait `std::fmt::Debug` is implemented for `Vec<T, A>`

```

</details>

danakj committed:
Avoid absolute sysroot paths in the MSVC linker command line

When the `--sysroot` is specified as relative to the current working
directory, the sysroot's rlibs should also be specified as relative
paths. Otherwise, the current working directory ends up in the
absolute paths, and in the linker command line. And the entire linker
command line appears in the PDB file generated by the MSVC linker.

When adding an rlib to the linker command line, if the rlib's canonical
path is in the sysroot's canonical path, then use the current sysroot
path + filename instead of the full absolute path to the rlib. This
means that when `--sysroot=foo` is specified, the linker command line
will contain `foo/rustlib/target/lib/lib*.rlib` instead of the full
absolute path to the same.

This addresses https://github.com/rust-lang/rust/issues/112586

danakj commented:
PTAL this is now done in link.rs and just rehomes rlib paths in the linker command line that are into the sysroot to use the actual sysroot path instead of the full absolute path.

danakj committed:
Move the comment on fixing paths to where it belongs

danakj commented:
Oh, can the stdlib crates by dylibs?

bjorn3 commented:
Yes, libstd.so and libtest.so exist. There are some (minor) limitations around their use though.

danakj committed:
Remap dylib paths into the sysroot to be relative to the sysroot

Like for rlibs, the paths on the linker command line need to be relative
paths if the sysroot was specified by the user to be a relative path.

Dylibs put the path in /LIBPATH instead of into the file path of the
library itself, so we rehome the libpath and adjust the rehoming function
to be able to support both use cases, rlibs and dylibs.

michaelwoerister commented:
Thanks, @danakj! This looks good to me. The next beta branch is still a while off, so we should find out in time if it has any unintended side effects. 

@bors r+ rollup=never

When we fix the issue with the object file in `/tmp`, I'd like us to look into improving the reproducibility tests and extend them to Windows.

bors commented:
:pushpin: Commit c340325ebf348914c67fc62e868db8833845eeb9 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister c340325ebf348914c67fc62e868db8833845eeb9 -->
<!-- homu: {"type":"Approved","sha":"c340325ebf348914c67fc62e868db8833845eeb9","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c340325ebf348914c67fc62e868db8833845eeb9 with merge 99b334696fffe8c08d2e6a978862849d5a89f875...
<!-- homu: {"type":"BuildStarted","head_sha":"c340325ebf348914c67fc62e868db8833845eeb9","merge_sha":"99b334696fffe8c08d2e6a978862849d5a89f875"} -->

rust-timer commented:
Finished benchmarking commit (99b334696fffe8c08d2e6a978862849d5a89f875): [comparison URL](https://perf.rust-lang.org/compare.html?start=0966f3202d1e811cd3aa35ac876b61a211b4819a&end=99b334696fffe8c08d2e6a978862849d5a89f875&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=0966f3202d1e811cd3aa35ac876b61a211b4819a&end=99b334696fffe8c08d2e6a978862849d5a89f875&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.6% | [1.6%, 1.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0966f3202d1e811cd3aa35ac876b61a211b4819a&end=99b334696fffe8c08d2e6a978862849d5a89f875&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6% | [2.6%, 2.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | 3.6% | [3.6%, 3.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.6% | [2.6%, 2.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0966f3202d1e811cd3aa35ac876b61a211b4819a&end=99b334696fffe8c08d2e6a978862849d5a89f875&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.3%  | [2.8%, 4.7%]   | 5     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.0% | [-5.7%, -0.9%] | 9     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.8% | [-5.7%, 4.7%]  | 14    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=0966f3202d1e811cd3aa35ac876b61a211b4819a&end=99b334696fffe8c08d2e6a978862849d5a89f875&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 648.963s -> 646.331s (-0.41%)
